---
author: "University of Melbourne Centre for Cancer Research"
date: "`r as.character(paste(Sys.time(), Sys.timezone()))`"
output:
  html_document:
    theme: flatly # cosmo # simplex
    toc: false
    code_download: true
    css: style.css
  rmdformats::material:
    highlight: kate
params:
  indir: "/path/to/dragen/results"
  outdir: "/path/to/output/dracarys"
  name: "SAMPLE_FOO"
title: "`r glue::glue('DRAGEN Results for {params$name}')`"
description: "Results from DRAGEN analysis"
---

<style type="text/css">
.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{r load_knitr, include=F}
require(knitr)
knitr::opts_chunk$set(
  echo = FALSE
)
```

```{r render_report_interactively, eval=F, echo=F}
params <- function(x) {
  paths <- list(
    local = list(
      indir = here::here("nogit/SEQCII/SEQCII_NT100"),
      outdir = here::here("nogit/SEQCII/SEQCII_NT100/dracarys"),
      name = "SEQCII_100")
  )
  paths[[x]]
}
params <- params("local")

render_me <- function() {
  rmarkdown::render(
    input = here::here("inst/rmd/tumor_normal/tumor_normal.Rmd"),
    params = params)
}
render_me()
```

```{r load_pkgs, message=F, warning=F}
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(dracarys))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(glue))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(jsonlite))
suppressPackageStartupMessages(library(kableExtra))
suppressPackageStartupMessages(library(patchwork))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(readr))
suppressPackageStartupMessages(library(tidyr))
```

```{r var_setup}
indir <- params$indir
outdir <- params$outdir
nm <- params$name

dracarys_env <- shQuote(Sys.getenv("DRACARYS_ENV"))
replay <- list.files(indir, pattern = "-replay.json", full.names = TRUE, recursive = TRUE)
v <- dracarys::setup_vars(replay)

eval_steps <- list(
  sv = v[[1]][["steps_run"]][["sv"]],
  varcall = v[[1]][["steps_run"]][["vc"]]
)
```

`r if (eval_steps$sv) { c("### Structural Variants {#structuralvars}") }`

```{r sv_prep, eval=eval_steps$sv, message=F, warning=F}
sv_fn <- v[[1]][["sv_fn"]]

circos_png <- dracarys::plot_circos(
  manta = sv_fn,
  env_path = dracarys_env,
  outdir = file.path(outdir, "circos")
)

sv_tab <- dracarys::sv_table(sv_fn)
sv_counts <- sv_tab$sv %>%
  dplyr::pull(SVTYPE) %>%
  table(useNA = "ifany") %>%
  addmargins() %>%
  tibble::as_tibble() %>%
  purrr::set_names(c("svtype", "count")) %>%
  dplyr::mutate(svtype = ifelse(svtype == "Sum", "Total",
                                ifelse(svtype == "BND", "BND (pairs)", svtype)))

```

```{r sv_results, eval=eval_steps$sv, out.width="40%"}
sv_counts %>%
  knitr::kable(format = "html", col.names = NULL, caption = "SV Type Counts") %>%
  kableExtra::kable_styling(full_width = FALSE, position = "float_left",
                            bootstrap_options = c("striped", "bordered", "condensed", "hover")) %>%
  kableExtra::row_spec(nrow(sv_counts), bold = T,
                       color = "white", background = "#2c3e50")

knitr::include_graphics(circos_png)
dracarys::plot_sv_contig_counts(sv_tab)

sv_tab$sv %>%
  dplyr::select(-num) %>%
  dplyr::select(chrom1, pos1, chrom2, pos2, svtype = SVTYPE,
                filter, score = SOMATICSCORE, imprecise = IMPRECISE,
                CIPOS, CIGAR, bnd_depth = BND_DEPTH, bnd_depth_mate = MATE_BND_DEPTH,
                dplyr::everything()) %>%
  dplyr::mutate_at(dplyr::vars(chrom1, chrom2, svtype, filter, imprecise), as.factor) %>%
  DT::datatable(rownames = FALSE,
                caption = "Structural Variants",
                filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                extensions = c("Scroller", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 400, scrollX = TRUE, keys = TRUE, dom = 'lfrtip')) %>%
  DT::formatCurrency(c("pos1", "pos2"),
                     currency = "", interval = 3, mark = ",", digits = 0)
```

`r if (eval_steps$varcall) { c("### Single Nucleotide/INDEL Variants {#snvars}") }`

```{r snv_prep, eval=eval_steps$snv}
snv_fn <- v[[1]][["snv_fn"]]
snv_tab <- dracarys::snv_table(snv_fn, prefix = nm, outdir = outdir, env_path = dracarys_env)
snv_samples <- snv_tab$samples

standard_chr <- paste0("chr", c(1:22, "X", "Y", "M"))
snv_vars <- snv_tab$variants %>%
  dplyr::rename(chr = CHROM, start = START, vartype = VARTYPE, filter = FILTER) %>%
  dplyr::mutate(filter = ifelse(filter == "None", "PASS", "FAIL"),
                filter = factor(filter, levels = c("PASS", "FAIL")),
                chr = ifelse(chr %in% standard_chr, chr, "ALT"),
                chr = factor(chr, levels = c(standard_chr, "ALT")))
```

```{r snv_results, eval=eval_steps$snv, fig.width=16, fig.height=8}
# summarise SNVs
snv_vars_summary <- snv_vars %>%
  dplyr::group_by(filter, vartype) %>%
  dplyr::count() %>%
  dplyr::ungroup() %>%
  dplyr::mutate_if(is.factor, as.character)

snv_vars_summary %>%
  dplyr::bind_rows(
    tibble::tribble(
      ~filter, ~vartype, ~n,
      "", "Total", sum(snv_vars_summary$n))) %>%
  knitr::kable(format = "html", col.names = NULL, caption = "SNV/INDEL Counts", format.args = list(big.mark = ",")) %>%
  kableExtra::kable_styling(full_width = FALSE, position = "float_left",
                            bootstrap_options = c("striped", "bordered", "condensed", "hover")) %>%
  kableExtra::row_spec(nrow(snv_vars_summary) + 1, bold = T,
                       color = "white", background = "#2c3e50")

snv_vars %>%
  dplyr::group_by(chr, filter, vartype) %>%
  dplyr::count() %>%
  ggplot2::ggplot(aes(x = chr, y = n, group = vartype)) +
  ggplot2::geom_point() +
  ggplot2::geom_line() +
  ggplot2::scale_x_discrete(drop = FALSE) +
  ggplot2::scale_y_continuous(
    breaks = scales::pretty_breaks(n = 10), limits = c(0, NA), expand = c(0, 0), labels = scales::comma) +
  ggplot2::theme_minimal() +
  ggplot2::labs(title = "Count of PASS/FAIL SNVs/INDELs per Chromosome") +
  ggplot2::xlab("Chromosome") +
  ggplot2::ylab("Count") +
  ggplot2::theme(
    panel.grid.minor = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(angle = 45, vjust = 1, hjust = 1, size = 10),
    plot.title = ggplot2::element_text(colour = "#2c3e50", size = 14, face = "bold"),
    panel.spacing = ggplot2::unit(2, "lines")) +
  facet_grid(filter~vartype, scales = "free")

# snv_tab_cap <- glue::glue("Single Nucleotide/INDEL Variants.\nSample1: {snv_samples$col1}; Sample2: {snv_samples$col2}")
```

### QC {.tabset .tabset-fade}

#### Coverage Plots
```{r qc_cov_contig_plot, fig.width=14, fig.height=10}
fine_hist <- dracarys::plot_wgs_fine_hist(xs = v[[1]][["cov_finehist_fn"]])
fine_cumsum <- dracarys::plot_wgs_fine_cumsum(xs = v[[1]][["cov_finehist_fn"]])
contig_cov <- dracarys::plot_wgs_contig_coverage(xs = v[[1]][["cov_contig_fn"]])

(fine_hist / fine_cumsum) | contig_cov
```

#### Runtimes
```{r qc_time_metrics}
time_metrics <- dracarys::read_time_metrics(v[[1]][["time_met_fn"]])

time_metrics %>%
  dplyr::select(-Label) %>%
  knitr::kable(format = "html", col.names = NULL, caption = "Time Metrics") %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left",
                            bootstrap_options = c("striped", "bordered",
                                                  "condensed", "hover")) %>%
  kableExtra::column_spec(1, bold = TRUE) %>%
  kableExtra::row_spec(nrow(time_metrics), bold = T,
                       color = "white", background = "#2c3e50")
```

#### Ploidy Estimation
```{r qc_ploidy_estimation_tab}
ploidy_estimation <- dracarys::read_ploidy_estimation_metrics(v[[1]][["ploidy_est_fn"]])
ploidy_estimation %>%
knitr::kable(format = "html", col.names = NULL, caption = "Ploidy Estimation") %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left",
                            bootstrap_options = c("striped", "bordered",
                                                  "condensed", "hover")) %>%
  kableExtra::column_spec(1, bold = TRUE) %>%
  kableExtra::row_spec(nrow(ploidy_estimation), bold = T,
                       color = "white", background = "#2c3e50")
```

#### Fragment Length Distribution
```{r qc_frag_len_distr, fig.height=3.5}
dracarys::plot_fragment_length_hist(v[[1]][["fraglen_fn"]])
```


#### Metrics Tables {.tabset .tabset-fade}

##### Coverage
```{r qc_cov_metrics}
cov_metrics <- purrr::map(v[[1]][["cov_met_fn"]], dracarys::read_wgs_coverage_metrics) %>%
 dplyr::bind_rows()
```

```{r qc_cov_metrics_count}
cov_metrics_count <- cov_metrics %>%
  tidyr::pivot_wider(id_cols = label,
                     names_from = var_abbrev,
                     values_from = count) %>%
  dplyr::select_if(colSums(!is.na(.)) > 0)

cov_metrics_count %>%
  DT::datatable(rownames = FALSE,
                # callback = JS(glue::glue("
                #   var tips = {tab_descr};
                #   var header = table.columns().header();
                #   for (var i = 0; i < tips.length; i++) {{
                #     $(header[i]).attr('title', tips[i]);
                #   }}")),
                caption = "Counts",
                class = "cell-border display compact",
                extensions = c("Scroller", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 150, keys = TRUE, dom = 'lfrtip')) %>%
  DT::formatCurrency(colnames(cov_metrics_count),
                     currency = "", interval = 3, mark = ",", digits = 0)
```

```{r qc_cov_metrics_pct}
cov_metrics_pct <- cov_metrics %>%
  tidyr::pivot_wider(id_cols = label,
                     names_from = var_abbrev,
                     values_from = pct) %>%
  dplyr::select_if(colSums(!is.na(.)) > 0)

cov_metrics_pct %>%
  DT::datatable(rownames = FALSE,
                # callback = JS(glue::glue("
                #   var tips = {tab_descr};
                #   var header = table.columns().header();
                #   for (var i = 0; i < tips.length; i++) {{
                #     $(header[i]).attr('title', tips[i]);
                #   }}")),
                caption = "Percentages",
                class = "cell-border display compact",
                extensions = c("Scroller", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 150,
                  keys = TRUE, dom = 'lfrtip'))
```

##### Mapping
```{r qc_map_metrics_tab}
mapping_metrics <- dracarys::read_mapping_metrics(v[[1]][["map_met_fn"]])
mapping_metrics <- list(
  data = mapping_metrics,
  abbrev = dplyr::distinct(mapping_metrics, var, var_abbrev),
  count = tidyr::pivot_wider(
    mapping_metrics, id_cols = c(Phenotype, RG),
    names_from = var_abbrev, values_from = count),
  pct = tidyr::pivot_wider(
    mapping_metrics, id_cols = c(Phenotype, RG),
    names_from = var_abbrev, values_from = pct)
)

mma <- mapping_metrics$abbrev
```

```{r qc_map_metrics_tab_pct}
dpct <- mapping_metrics$pct %>%
  dplyr::select_if(colSums(!is.na(.)) > 0)
tab_descr <- mma$var[match(colnames(dpct), mma$var_abbrev)] %>%
  jsonlite::toJSON()

dpct %>%
  DT::datatable(rownames = FALSE,
                callback = JS(glue::glue("
                  var tips = {tab_descr};
                  var header = table.columns().header();
                  for (var i = 0; i < tips.length; i++) {{
                    $(header[i]).attr('title', tips[i]);
                  }}")),
                caption = "Percentages. Bar range: 90-100",
                class = "cell-border display compact",
                extensions = c("Scroller", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 150, scrollX = TRUE,
                  autoWidth = TRUE, keys = TRUE, dom = 'lfrtip')) %>%
  DT::formatStyle(colnames(dpct),
                  background = DT::styleColorBar(c(90, 100), color = "lightblue"),
                  backgroundSize = "90% 90%", backgroundRepeat = 'no-repeat',
                  backgroundPosition = 'center')
```

```{r qc_map_metrics_tab_count}
dcount <- mapping_metrics$count %>%
  dplyr::select_if(colSums(!is.na(.)) > 0)
tab_descr <- mma$var[match(colnames(dcount), mma$var_abbrev)] %>%
  jsonlite::toJSON()

dcount %>%
  DT::datatable(rownames = FALSE,
                callback = JS(glue::glue("
                  var tips = {tab_descr};
                  var header = table.columns().header();
                  for (var i = 0; i < tips.length; i++) {{
                    $(header[i]).attr('title', tips[i]);
                  }}")),
                caption = "Counts.",
                class = "cell-border display compact",
                extensions = c("Scroller", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 150, scrollX = TRUE,
                  autoWidth = TRUE, keys = TRUE, dom = 'lfrtip')) %>%
  DT::formatCurrency(colnames(dcount),
                     currency = "", interval = 3, mark = ",", digits = 0)
```

`r if (eval_steps$varcall) { c("##### VarCalling") }`

```{r qc_vc_metrics_tab, eval=eval_steps$varcall}
vc_metrics <- dracarys::read_varcalling_metrics(v[[1]][["vc_met_fn"]]) %>%
  dplyr::filter(category != "summary")
vc_metrics <- list(
  data = vc_metrics,
  count = tidyr::pivot_wider(
    vc_metrics, id_cols = c(category, sample),
    names_from = var, values_from = count),
  pct = tidyr::pivot_wider(
    vc_metrics, id_cols = c(category, sample),
    names_from = var, values_from = pct)
)
```

```{r qc_vc_metrics_tab_count, eval=eval_steps$varcall}
dcount <- vc_metrics$count %>%
  dplyr::select_if(colSums(!is.na(.)) > 0)
# tab_descr <- mma$var[match(colnames(dcount), mma$var_abbrev)] %>%
#   jsonlite::toJSON()

dcount %>%
  DT::datatable(rownames = FALSE,
                # callback = JS(glue::glue("
                #   var tips = {tab_descr};
                #   var header = table.columns().header();
                #   for (var i = 0; i < tips.length; i++) {{
                #     $(header[i]).attr('title', tips[i]);
                #   }}")),
                caption = "Counts.",
                class = "cell-border display compact",
                extensions = c("Scroller", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 150, scrollX = TRUE,
                  autoWidth = TRUE, keys = TRUE, dom = 'lfrtip')) %>%
  DT::formatCurrency(colnames(dcount),
                     currency = "", interval = 3, mark = ",", digits = 0)
```

```{r qc_vc_metrics_tab_pct, eval=eval_steps$varcall}
dpct <- vc_metrics$pct %>%
  dplyr::select_if(colSums(!is.na(.)) > 0)
# tab_descr <- mma$var[match(colnames(dpct), mma$var_abbrev)] %>%
#   jsonlite::toJSON()

dpct %>%
  DT::datatable(rownames = FALSE,
                # callback = JS(glue::glue("
                #   var tips = {tab_descr};
                #   var header = table.columns().header();
                #   for (var i = 0; i < tips.length; i++) {{
                #     $(header[i]).attr('title', tips[i]);
                #   }}")),
                caption = "Percentages. Bar range: 90-100",
                class = "cell-border display compact",
                extensions = c("Scroller", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 150,
                  autoWidth = TRUE, keys = TRUE, dom = 'lfrtip')) %>%
  DT::formatStyle(colnames(dpct),
                  background = DT::styleColorBar(c(90, 100), color = "lightblue"),
                  backgroundSize = "90% 90%", backgroundRepeat = 'no-repeat',
                  backgroundPosition = 'center')
```

#### {-}

## Addendum {.tabset .tabset-fade #addendum}

### DRAGEN {.tabset .tabset-fade}

#### Version
```{r dragen_replay_system}
replay <- dracarys::read_replay(v[[1]][["replay_fn"]])
replay$system %>%
  knitr::kable(format = "html", col.names = NULL) %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left",
                            bootstrap_options = c("striped", "bordered",
                                                  "condensed", "hover")) %>%
  kableExtra::column_spec(1, bold = TRUE)
```

#### Parameters
```{r dragen_replay_config}
replay$dragen_config %>%
  DT::datatable(rownames = FALSE,
                filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                extensions = c("Scroller", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 150,
                  autoWidth = FALSE, keys = TRUE, dom = 'lfrtip'))
```

### Report Inputs
```{r report_inputs, eval=T}
dplyr::tibble(name = names(params), value = unlist(params)) %>%
  knitr::kable(format = "html") %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left",
                            bootstrap_options = c("striped", "bordered",
                                                  "condensed", "hover")) %>%
  kableExtra::column_spec(1, bold = TRUE)
```

### R Session Info
```{r session_info, eval=T}
si <- devtools::session_info(include_base = TRUE)
unclass(si$packages) %>%
  dplyr::as_tibble() %>%
  dplyr::mutate(main_pkgs = ifelse(package %in% c("woofr", "rock", "base"), "A", "B")) %>%
  dplyr::arrange(main_pkgs, package) %>%
  dplyr::select(package, version = ondiskversion, path, datestamp = date) %>%
  knitr::kable(format = "html") %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left",
                            bootstrap_options = c("striped", "bordered",
                                                  "condensed", "hover")) %>%
  kableExtra::column_spec(1, bold = TRUE) %>%
  kableExtra::scroll_box(height = "200px")
```

***

<p>&nbsp;</p>
